File: Codes/Level 2/Recursion and Backtracking/Combination Sum Problems/CombinationSum1.cpp
class Solution {
public:
    
    map<vector<int>, int> m;
    
    void combine(vector<int> &candidates, vector<vector<int>> &ans, vector<int> currAns, int currSum, int target){
        if(currSum == target){
            sort(currAns.begin(), currAns.end());
            if(m.find(currAns) == m.end()){
                ans.push_back(currAns);
                m[currAns]++;
            }
            return;
        }
        for(int i=0; i<candidates.size(); i++){
            if(currSum + candidates[i] <= target){
                currAns.push_back(candidates[i]);
                combine(candidates, ans, currAns, currSum + candidates[i], target);
                currAns.pop_back();
            }
        }
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> currAns;
        combine(candidates, ans, currAns, 0, target);
        return ans;
    }
};

File: Codes/Level 2/Recursion and Backtracking/Combination Sum Problems/CombinationSum1.java
class Solution {
    
    HashMap<List<Integer>, Integer> unique;
    
    public void combine(int[] candidates, int target, List<Integer> currAns, List<List<Integer>> ans){
        if(target == 0){
            List<Integer> temp = new ArrayList<>();
            for(Integer i : currAns)    temp.add(i);
            Collections.sort(temp);
            if(unique.containsKey(temp) == false){
                ans.add(temp);
                unique.put(temp, 1);
            }
            return;
        }
        
        for(int i=0; i<candidates.length; i++){
            if(target - candidates[i] >= 0){
                currAns.add(candidates[i]);
                combine(candidates, target - candidates[i], currAns, ans);
                currAns.remove(currAns.size() - 1);
            }
        }
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> currAns = new ArrayList<>();
        unique = new HashMap<>();
        combine(candidates, target, currAns, ans);
        return ans;
    }
}